nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp

```diff
@@ -56,24 +56,18 @@ class Layer  // TODO(mjeronimo): public nav2_lifecycle::LifecycleHelperInterface
   Layer();
   virtual ~Layer() {}
 
+  // TODO(mjeronimo): should the following functions changed to a lifecycle-style interface?
   void initialize(
     LayeredCostmap * parent,
     std::string name,
     tf2_ros::Buffer * tf,
     nav2_lifecycle::LifecycleNode::SharedPtr node,
-    rclcpp::Node::SharedPtr client_node);
+    rclcpp::Node::SharedPtr client_node,
+    rclcpp::Node::SharedPtr rclcpp_node);
```

Comment:
Can you clarify why we need to pass in 3 node pointers for each layer initialization?

Reply:
Sure. Issues with ROS2. There is the main (lifecycle) node. This is where a helper, such as a plugin, can find node parameters, or use this node to create pub/subs, etc. There are still classes in ROS2 that don't properly support lifecycle nodes. So, the rclcpp_node is a temporary bridge to interface with these classes. rclcpp_node is a node that is spinning on an executor. Finally, the client_node is a node that can be used in a service call; it cannot be on an executor as ROS2 spins this node in the service call and will fail if the node is already spinning. 

I'm working to reduce this to a single node. This will require further changes in ROS2 to update those classes that don't yet support lifecycle nodes and also address the inability to use an already spinning node in a service call.